rfcs/20230622-quantized-reduction.md

```diff
@@ -0,0 +1,521 @@
+         tensor<... x !quant.uniform<i32:f32, input_scale:0>>)
+    -> tensor<... x !quant.uniform<i32:f32, input_scale:0>>
+```
+
+In this option, the `init_value` type and the `result` type can be different
+from the input type. The first argument of the compute block is fixed for the
+traversed element and the second argument is fixed for the intermediate
+(accumulation) result.
+
+#### Tradeoffs
+
+* (+) Make the accumulation type explicit in the IR.
+* (-) This representation imposes a limitation on the evaluation order.
+  Since we can’t express the computation between two intermediate (accumulation)
+  results, we can not arbitrarily insert `init_value` and start the
+  computation at an arbitrary location. The following shows the restricted
+  evaluation order with the method.
+
+```python
++----------+       +--------+ +--------+ +--------+ +--------+
+|init_value|       |input[0]| |input[1]| |input[2]| |input[3]|
++----------+       +--------+ +--------+ +--------+ +--------+
+           \        /         /          /          /
+           +-------+         /          /          /
+           |compute|        /          /          /
+           +-------+       /          /          /
+                \         /          /          /
+                 +-------+          /          /
+                 |compute|         /          /
+                 +-------+        /          /
+                       \         /          /
+                        +-------+          /
+                        |compute|         /
+                        +-------+        /
+                              \         /
+                               +-------+
+                               |compute|
+                               +-------+
+```
+
+## Open Question
+
+### Should we restrict the proposal #1 to quantized types only?
+
+The above proposal #1 of introducing the additional functions is theoretically
+not limited to quantized `reduce` op, but also can be applied to `reduce` op with
+non-quantized types. For example,
+
+```mlir
+%result = "stablehlo.reduce"(%input, %init_value) ({
+  ^input_conversion(%arg0: tensor<bf16>):
+    %0 = "stablehlo.convert"(%arg0): (tensor<bf16>) -> (tensor<f32>)
+    "stablehlo.return"(%0) : (tensor<f32>) -> (tensor<f32>)
+  }, {
+  ^bb0(%arg0: tensor<f32>, %arg1: tensor<f32>):
+    %0 = "stablehlo.add"(%arg0, %arg1) : (tensor<f32>, tensor<f32>) ->
+    tensor<f32>
+    "stablehlo.return"(%0) : (tensor<f32>) -> ()
+  },
+  ^output_conversion(%arg0: tensor<f32>):
+    %0 = "stablehlo.convert"(%arg0): (tensor<f32>) -> (tensor<bf16>)
+    "stablehlo.return"(%0) : (tensor<bf16>) -> (tensor<bf16>)
+  }) {
+  dimensions = dense<1> : tensor<1xbf16>
+} : (tensor<1x6xbf16>, tensor<bf16>) -> tensor<1xbf16>
+```
+
+However, it is not clear how such operations will be lowered to other IR
+representations, like HLO, which does not support such additional computation
```

Comment:
That is the problem of HLO.  Either they decide not to support it and reject the op or extend the functionality of HLO.

The existing backend do 16-bit floating point work by either pattern matching, semantic/algorithmetic rewrite, use `bf16` to wish its range is large enough, or just guessing the intention of the programmer because we don't have a good representation.  I really hope we can include this case.

This design makes sense to CPU or GPU like architecture when they have bf16 to f32 extension internally.

Reply:
noted.